### PR TITLE
bring const to RootObject.equals()

### PR DESCRIPTION
**Pull request by WalterBright**

### CODE REVIEW

**Comment by thewilsonator** [comment]
Semaphore:
```
dmd/identifier.d:85:19: error: function dmd.identifier.Identifier.equals does not override any function, did you mean to override 'extern (C++) dmd.root.rootobject.RootObject.equals'?
     override bool equals(const RootObject o) const
                   ^
```

**Comment by WalterBright** [comment]
It's a strange error. It should compile everywhere else, and does.

File: src/dmd/func.d [hunk 1]
```diff
@@ -526,29 +526,38 @@ extern (C++) class FuncDeclaration : Declaration
...
         {
-            alias fd1 = this;
-            auto  fd2 = s.isFuncDeclaration();
+            auto fd1 = this;
+            auto fd2 = s.isFuncDeclaration();
             if (!fd2)
                 return false;
 
             auto fa1 = fd1.isFuncAliasDeclaration();
+	    auto faf1 = fa1 ? fa1.toAliasFunc() : fd1;
```
Comment (jacob-carlborg) [inline src/dmd/func.d]: Indentation is off.

File: src/dmd/func.d [hunk 2]
```diff
@@ -526,29 +526,38 @@ extern (C++) class FuncDeclaration : Declaration
...
-            if (fa1 && (fd1 = fa1.toAliasFunc()).isUnique() && !fa1.hasOverloads)
-                fa1 = null;
-            if (fa2 && (fd2 = fa2.toAliasFunc()).isUnique() && !fa2.hasOverloads)
-                fa2 = null;
-            if ((fa1 !is null) != (fa2 !is null))
+	    bool b1 = fa1 !is null;
+            if (b1 && faf1.isUnique() && !fa1.hasOverloads)
+                b1 = false;
+
+	    bool b2 = fa2 !is null;
```
Comment (jacob-carlborg) [inline src/dmd/func.d]: Indentation if off.

**Review by jacob-carlborg — CHANGES_REQUESTED** [review]

**Comment by WalterBright** [comment]
I've failed again at making `equals()` const. But these changes move things along a bit.

The stumbling block is `denum.EnumDeclaration.getMemtype()` resolving a forward reference to an enum.

**Comment by WalterBright** [comment]
`overloadApply()` is another problem, because it prints error messages, which modify the dsymbol by setting `errors`.

**Review by jacob-carlborg — DISMISSED** [review]

**Review by wilzbach — APPROVED** [review]

**Pull request merged**